forge-game/src/main/java/forge/game/card/Card.java

```diff
@@ -4327,6 +4327,42 @@ public final boolean hasIntensity() {
         return intensity > 0;
     }
 
+    private Map<String, List<Object>> perpetual = Maps.newHashMap();
+    public final boolean hasPerpetual() {
+        return !perpetual.isEmpty();
+    }
+    public final Map<String, List<Object>> getPerpetual() {
+        return perpetual;
+    }
+
+    public final void addPerpetual(String s, List<Object> list) {
+        perpetual.put(s, list);
+        executePerpetual(s, list);
+    }
+
+    public final void executePerpetual(String s, List<Object> list) {
+        if (s.equals("NewPT")) {
+            addNewPT((Integer) list.get(0), (Integer) list.get(1), (long) list.get(2), (long) list.get(3));
+        } else if (s.equals("PTBoost")) {
+            addPTBoost((Integer) list.get(0), (Integer) list.get(1), (long) list.get(2), (long) list.get(3));
+        }
+    }
+
+    public final void setPerpetual(Map<String, List<Object>> map) {
+        for (String key : map.keySet()) {
+            addPerpetual(key, map.get(key));
+        }
+    }
+
+    public final void generatePerpetual(String type, Integer power, Integer toughness, long timestamp) {
```

Comment:
I think `String type, long timestamp, Object... params` so it can be reused by all effects

Reply:
I reworked this a fair amount